### ISSUE TITLE
Multi-fields search in documents is not working

### ISSUE THREAD

**Issue by dfanchon** [nextapps-de/flexsearch#264]
Hi,

Following the doc this sample code should work but it yields an empty array.
Any clues about what is done wrong?

```
var { Document } = require(`flexsearch`);

var index = new Document({
  document: {
    id: "id",
    index: ["title"]
  }
});

index.add({
  id: 0,
  title: "I am a test"
});
index.add({
  id: 1,
  title: "I am also a test"
});

var result = index.search([
  {
    field: "title",
    query: "test"
  },
  {
    field: "title",
    query: "also"
  }
]);
console.log(`result`, JSON.stringify(result));
```

**Comment by zuramai** [comment]
same here

**Comment by theguy147** [comment]
same issue here. 

I had a quick look at the sources and I think the issue is: https://github.com/nextapps-de/flexsearch/blob/65b027ca316ceefd8ea89f472561a0f91179b2f3/src/index.js#L444-L447

when doing a multi-field search on a document flexsearch appears to call search on each index without setting `query`.  if `query` were set then `length` would be assigned in this block of code:  https://github.com/nextapps-de/flexsearch/blob/master/src/index.js#L405-L442

But because that block is not being executed `length` is not set and therefore `search` returns early (without results).

(I haven't checked my findings yet because I don't currently have a device nearby that I can use for a testing environment)

If I'm right either `Document.search` needs to retrieve `query` from `opt` and pass it along or this has to be done inside `Index.search` if no `query` was provided.

**Comment by theguy147** [comment]
Ok, finally got around to testing it and this fixes it for me:

https://github.com/nextapps-de/flexsearch/blob/65b027ca316ceefd8ea89f472561a0f91179b2f3/src/index.js#L397-L403

This block could to be changed to include e.g.:

```javascript
query = query || options["query"];
```

**Comment by theguy147** [comment]
The source of the issue is L481 in `document.js`.`query` remains `undefined` in the example used by @dfanchon  because `options["query"]` is not defined in that case.

https://github.com/nextapps-de/flexsearch/blob/65b027ca316ceefd8ea89f472561a0f91179b2f3/src/document.js#L481


### Fix ###

This would be a fix targeting `document.js` directly instead of the fix i suggested above in `index.js`:

```diff
diff --git a/src/document.js b/src/document.js
index 25da47c..f708651 100644
--- a/src/document.js
+++ b/src/document.js
@@ -110,7 +110,8 @@ function parse_descriptor(options, document){
         if(!is_string(key)){
 
             opt = key;
-            key = key["field"];
+            key = opt["field"];
+            query = query || opt["query"];
         }
 
         opt = is_object(opt) ? Object.assign({}, options, opt) : options;
```

**Comment by theguy147** [comment]
ups, nevermind. my second fix doesn't work as intended because in the second iteration of the loop `query` has already been set and remains the same for all remaining iterations.

I guess it could be solved by either introducing a new variable or something like this:

```diff
diff --git a/src/document.js b/src/document.js
index 25da47c..5fd44c5 100644
--- a/src/document.js
+++ b/src/document.js
@@ -563,7 +563,7 @@ Document.prototype.search = function(query, limit, options, _resolve){
 
         if(promises){
 
-            promises[i] = this.index[key].searchAsync(query, limit, opt || options);
+            promises[i] = this.index[key].searchAsync((opt && opt["query"]) || query, limit, opt || options);
 
             // just collect and continue
 
@@ -577,7 +577,7 @@ Document.prototype.search = function(query, limit, options, _resolve){
 
             // inherit options also when search? it is just for laziness, Object.assign() has a cost
 
-            res = this.index[key].search(query, limit, opt || options);
+            res = this.index[key].search((opt && opt["query"]) || query, limit, opt || options);
         }
 
         len = res && res.length;
```

but performance-wise both of these solutions might not be optimal. so for now I will stick to my first suggestion (to modify `index.js`).

**Comment by GanserITService** [comment]
@theguy147 when is the next release planned with the fix?

**Comment by micheljung** [comment]
@theguy147 do you have any idea how this works with `{enrich: true}`? I feel like there's another bug :)

**Comment by a-tonchev** [comment]
I also face this bug, what would be the workaround?

**Comment by mmm8955405** [comment]
I also face this bug, what would be the workaround?

**Comment by mmm8955405** [comment]
I think the stable production version should not make such low-level mistakes

**Comment by ts-thomas** [comment]
Automated tests will be back in the next version. Sorry for the circumstances.

**Comment by ts-thomas** [comment]
Additional fix was added in v0.7.23
